src/blueprint.zig

```diff
@@ -0,0 +1,281 @@
+				const worldY = startY + @as(i32, @intCast(offsetY));
+
+				for(0..self.sizeZ) |offsetZ| {
+					const worldZ = startZ + @as(i32, @intCast(offsetZ));
+
+					const block = self.blocks.items[blockIndex];
+					_ = main.server.world.?.updateBlock(worldX, worldY, worldZ, block);
+
+					blockIndex += 1;
+				}
+			}
+		}
+	}
+	pub fn store(self: @This(), allocator: NeverFailingAllocator) []u8 {
+		std.debug.assert(self.sizeX != 0);
+		std.debug.assert(self.sizeY != 0);
+		std.debug.assert(self.sizeZ != 0);
+
+		var gameIdToBlueprintId = self.makeGameIdToBlueprintIdMap(main.stackAllocator);
+		defer gameIdToBlueprintId.deinit();
+		std.debug.assert(gameIdToBlueprintId.count() != 0);
+
+		const blockPalette = self.packBlockPalette(main.stackAllocator, gameIdToBlueprintId);
+		defer main.stackAllocator.free(blockPalette);
+		std.debug.assert(gameIdToBlueprintId.count() == blockPalette.len);
+
+		const blockPaletteSizeBytes = self.getBlockPaletteSize(blockPalette);
+		const blockArraySizeBytes = self.sizeX*self.sizeY*self.sizeZ*@sizeOf(BlockStorageType);
+		const decompressedDataSizeBytes = blockPaletteSizeBytes + blockArraySizeBytes;
+		std.debug.assert(decompressedDataSizeBytes != 0);
+
+		var decompressedWriter = BinaryWriter.initCapacity(allocator, .big, decompressedDataSizeBytes);
+		defer decompressedWriter.deinit();
+
+		self.storeBlockPalette(&decompressedWriter, blockPalette);
+		self.storeBlockArray(&decompressedWriter, gameIdToBlueprintId);
+
+		const compression: BlueprintCompression = .deflate;
+		const compressedData: []u8 = self.compressOutputBuffer(main.stackAllocator, decompressedWriter.data.items, compression);
+		defer main.stackAllocator.free(compressedData);
+		std.debug.assert(compressedData.len != 0);
+
+		var outputWriter = BinaryWriter.initCapacity(allocator, .big, @sizeOf(i16) + @sizeOf(BlueprintCompression) + @sizeOf(u32) + @sizeOf(u16)*4 + compressedData.len);
+
+		outputWriter.writeInt(u16, blueprintVersion);
+		outputWriter.writeEnum(BlueprintCompression, compression);
+		outputWriter.writeInt(u32, @intCast(blockPaletteSizeBytes));
+		outputWriter.writeInt(u16, @intCast(gameIdToBlueprintId.count()));
+		outputWriter.writeInt(u16, @intCast(self.sizeX));
+		outputWriter.writeInt(u16, @intCast(self.sizeY));
+		outputWriter.writeInt(u16, @intCast(self.sizeZ));
+
+		outputWriter.writeSlice(compressedData);
+
+		return outputWriter.data.toOwnedSlice();
+	}
+	fn makeGameIdToBlueprintIdMap(self: @This(), allocator: NeverFailingAllocator) GameIdToBlueprintIdMapType {
+		var gameIdToBlueprintId: GameIdToBlueprintIdMapType = .init(allocator.allocator);
+
+		for(self.blocks.items) |block| {
+			const result = gameIdToBlueprintId.getOrPut(block.typ) catch unreachable;
+			if(!result.found_existing) {
+				result.value_ptr.* = @intCast(gameIdToBlueprintId.count() - 1);
+			}
+		}
+
+		return gameIdToBlueprintId;
+	}
+	fn packBlockPalette(_: @This(), allocator: NeverFailingAllocator, map: GameIdToBlueprintIdMapType) [][]const u8 {
```

Comment:
if you don't use self, then you shouldn't take it as a parameter in my opinion.

Reply:
Probably, I'm used to Python where it is always present unless marked `@staticmethod` or `@classmethod`.